<a href="https://colab.research.google.com/github/Sudesh8/Recipe_Recommandaation/blob/main/nashik_Project_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [61]:
import pandas as pd
import numpy as np

In [3]:
recipe_df=pd.read_csv('/content/project_food.csv')

In [4]:
recipe_df.head()

recipe_id                               TranslatedRecipeName  \
0          0                               Masala Karela Recipe   
1          1                         Spicy Tomato Rice (Recipe)   
2          2  Ragi Semiya Upma Recipe - Ragi Millet Vermicel...   
3          3  Gongura Chicken Curry Recipe - Andhra Style Go...   
4          4  Andhra Style Alam Pachadi Recipe - Adrak Chutn...   

                               TranslatedIngredients  TotalTimeInMins  \
0  1 tablespoon Red Chilli powder,3 tablespoon Gr...               45   
1   2 teaspoon cashew - or peanuts, 1/2 Teaspoon ...               15   
2  1 Onion - sliced,1 teaspoon White Urad Dal (Sp...               50   
3  1/2 teaspoon Turmeric powder (Haldi),1 tablesp...               45   
4   oil - as per use, 1 tablespoon coriander seed...               30   

                Cuisine                             TranslatedInstructions  \
0                Indian  To begin making the Masala Karela Recipe,de-se...   
1  South Indian Recipes  To make tomato puliogere, first cut the tomato...   
2  South Indian Recipes  To begin making the Ragi Vermicelli Recipe, fi...   
3                Andhra  To begin making Gongura Chicken Curry Recipe f...   
4                Andhra  To make Andhra Style Alam Pachadi, first heat ...   

                                 Cleaned-Ingredients  Ingredient-count  
0  salt,amchur (dry mango powder),karela (bitter ...                10  
1  tomato,salt,chickpea lentils,green chilli,rice...                12  
2  salt,rice vermicelli noodles (thin),asafoetida...                12  
3  tomato,salt,ginger,sorrel leaves (gongura),fen...                15  
4  tomato,salt,ginger,red chillies,curry,asafoeti...                12

In [5]:
user_df=pd.read_csv('/content/user_activitY (1).csv')

In [7]:
user_df.head()

id  user_id  recipe_id  action            timestamp
0   1        3       5048  search  2025-01-03 11:49:07
1   2        3       2491  search  2025-01-03 04:03:46
2   3        2       2438  search  2025-01-03 10:20:30
3   4        1       1219  search  2025-01-02 11:47:53
4   5        3       3887    view  2025-01-02 10:29:47

In [8]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity


In [9]:
# TF-IDF Vectorization
vectorizer = TfidfVectorizer()

In [10]:

ingredient_vectors = vectorizer.fit_transform(recipe_df["Cleaned-Ingredients"])

In [11]:
feature_names = vectorizer.get_feature_names_out()  # Get feature names (vocabulary)
tfidf_df = pd.DataFrame(ingredient_vectors.toarray(), columns=feature_names)


In [12]:
tfidf_df

aa  aachi  aam  aamras  aar  aata  above  access  accroding  achar  \
0     0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
1     0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
2     0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
3     0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
4     0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
...   ...    ...  ...     ...  ...   ...    ...     ...        ...    ...   
5933  0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
5934  0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
5935  0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
5936  0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   
5937  0.0    0.0  0.0     0.0  0.0   0.0    0.0     0.0        0.0    0.0   

      ...  yogurt  you  young   za  zest  ziti  zucchini  zucchinilengthwise  \
0     ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
1     ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
2     ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
3     ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
4     ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
...   ...     ...  ...    ...  ...   ...   ...       ...                 ...   
5933  ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
5934  ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
5935  ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
5936  ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   
5937  ...     0.0  0.0    0.0  0.0   0.0   0.0       0.0                 0.0   

      zucchinisesame  zucchiniwith  
0                0.0           0.0  
1                0.0           0.0  
2                0.0           0.0  
3                0.0           0.0  
4                0.0           0.0  
...              ...           ...  
5933             0.0           0.0  
5934             0.0           0.0  
5935             0.0           0.0  
5936             0.0           0.0  
5937             0.0           0.0  

[5938 rows x 2385 columns]

In [13]:
ingredient_vectors.toarray()

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [14]:
user_df.head(1)

id  user_id  recipe_id  action            timestamp
0   1        3       5048  search  2025-01-03 11:49:07

In [15]:
recipe_df.head(1)

recipe_id  TranslatedRecipeName  \
0          0  Masala Karela Recipe   

                               TranslatedIngredients  TotalTimeInMins Cuisine  \
0  1 tablespoon Red Chilli powder,3 tablespoon Gr...               45  Indian   

                              TranslatedInstructions  \
0  To begin making the Masala Karela Recipe,de-se...   

                                 Cleaned-Ingredients  Ingredient-count  
0  salt,amchur (dry mango powder),karela (bitter ...                10

In [16]:
user_activity = user_df[user_df["user_id"] == 1]


In [17]:
user_activity

id  user_id  recipe_id  action            timestamp
3    4        1       1219  search  2025-01-02 11:47:53
5    6        1       3088    view  2025-01-03 12:04:07
6    7        1       4889    like  2025-01-03 06:52:31
7    8        1        685    like  2025-01-03 08:08:57
8    9        1       1267    view  2025-01-02 19:20:13
11  12        1       5877  search  2025-01-02 22:30:54
14  15        1        466  search  2025-01-01 21:02:54
16  17        1        929    like  2025-01-02 06:24:24
22  23        1       4281  search  2025-01-02 22:04:18
25  26        1       1087    like  2025-01-02 06:52:57
27  28        1       1332    like  2025-01-02 04:13:28
28  29        1        684    like  2025-01-02 04:36:44
32  33        1       2001    view  2025-01-02 20:21:56
34  35        1       4557    view  2025-01-03 07:10:26
35  36        1       4417  search  2025-01-03 05:42:53
39  40        1       5354  search  2025-01-02 08:01:28
40  41        1       2262    like  2025-01-03 11:35:22
42  43        1       1803  search  2025-01-02 00:53:26
44  45        1       4157    like  2025-01-02 05:13:59
45  46        1       2271  search  2025-01-01 20:48:51

In [18]:
 # Filter recipes based on user history
recipe_ids = user_activity["recipe_id"].tolist()


In [19]:
type(recipe_ids),recipe_ids

(list,
 [1219,
  3088,
  4889,
  685,
  1267,
  5877,
  466,
  929,
  4281,
  1087,
  1332,
  684,
  2001,
  4557,
  4417,
  5354,
  2262,
  1803,
  4157,
  2271])

In [20]:
user_recipes = recipe_df[recipe_df["recipe_id"].isin(recipe_ids)]

In [21]:
user_recipes

recipe_id                               TranslatedRecipeName  \
466         466                     Szechuan Vegetarian Fried Rice   
684         684  Kachi Tikya Recipe  - Chicken In Tomato Gravy ...   
685         685  Nei Appam / Unniyappam Recipe - Coconut Banana...   
929         929  Andhra Style Mirapakaya Bajji Recipe - Stuffed...   
1087       1087                      Mumbai Style Pani Puri Recipe   
1219       1219  Arbi Ke Kebab Recipe (Spicy Colocasia Kebab) R...   
1267       1267  Chickpea Tamarind Broth Recipe - Tolerant Sout...   
1332       1332                            Baked Bhakarwadi Recipe   
1803       1803                Mutton KM Samos Recipe-Meat Singara   
2001       2001        Spicy Potato Roast Recipe (Recipe In Hindi)   
2262       2262                Cheesy Corn And Onion Toasts Recipe   
2271       2271  Karnataka Style Bassaru Recipe - Dill flavored...   
3088       3088                          Cardamom Macaroons Recipe   
4157       4157         Zucchini And Feta Cheese Bruschetta Recipe   
4281       4281                        Sauteed Spicy Prawns Recipe   
4417       4417           Rajma Aloo Cutlet Recipe (Tikki / Patty)   
4557       4557              Greek Style Grilled Pita Pizza Recipe   
4889       4889               Vegetarian Mexican Fried Rice Recipe   
5354       5354                              Tomato Chutney Recipe   
5877       5877      Kollu Rasam Recipe - Healthy Horse Gram Rasam   

                                  TranslatedIngredients  TotalTimeInMins  \
466   1 Carrot (Gajjar) - finely chopped,Salt - to t...               50   
684   4 Tomatoes - finely chopped,1 kg Chicken - min...               46   
685   3/4 cup Jaggery - powdered,2 teaspoons Cardamo...              220   
929   Sunflower Oil - for deep frying,1 teaspoon Cum...               40   
1087  1/4 Boondi,2 cups Sweet Chutney (Date & Tamari...               30   
1219  1 Green Chilli,350 grams Colocasia root (Arbi)...               70   
1267  1/2 teaspoon Mustard seeds,3 teaspoons Sambar ...               45   
1332  1 teaspoon Cumin powder (Jeera),Water - to kne...               60   
1803  1 teaspoon Kashmiri Red Chilli Powder,2 teaspo...               60   
2001   1/2 teaspoon Asafoetida, 6 Curry leaves, 1/2 ...               40   
2262  Salt - to taste,1 cup Sweet corn - boiled,1/2 ...               10   
2271  4 Dry Red Chillies,4 cloves Garlic,2 sprig Dil...               55   
3088  2 Egg whites - aged for 2-3 days at room tempe...               70   
4157  2 tablespoon Extra Virgin Olive Oil,1/2 teaspo...               25   
4281  1 teaspoon Turmeric powder (Haldi),3 sprig Cor...               35   
4417  1/2 tablespoon Amchur (Dry Mango Powder),1 tab...               60   
4557  5 Black olives,1 cup Homemade Hummus,4 Mini Pi...               30   
4889  4 cloves Garlic - minced,Salt - to taste,2 cup...               40   
5354  Salt,1 Dry Red Chilli,Sunflower Oil - as requi...               35   
5877  2 Tomatoes - made into a puree,1 teaspoon Cori...               70   

                    Cuisine  \
466                 Sichuan   
684    North Indian Recipes   
685          Kerala Recipes   
929                  Andhra   
1087   North Indian Recipes   
1219                 Awadhi   
1267   South Indian Recipes   
1332  Maharashtrian Recipes   
1803                 Indian   
2001   South Indian Recipes   
2262        Italian Recipes   
2271              Karnataka   
3088                 French   
4157        Italian Recipes   
4281                 Indian   
4417                 Indian   
4557                  Greek   
4889                Mexican   
5354   South Indian Recipes   
5877   South Indian Recipes   

                                 TranslatedInstructions  \
466   We begin making the Szechuan Vegetable Fried R...   
684   To begin making the Kachi Tikya recipe, marina...   
685   To begin making the Nei Appam / Unniyappam rec...   
929   To begin making the Andhra Style Mirapakaya 

In [33]:
y=user_recipes['recipe_id']

In [39]:
y.to_list()

[466,
 684,
 685,
 929,
 1087,
 1219,
 1267,
 1332,
 1803,
 2001,
 2262,
 2271,
 3088,
 4157,
 4281,
 4417,
 4557,
 4889,
 5354,
 5877]

In [25]:
x=recipe_ids

In [32]:
x

[1219,
 3088,
 4889,
 685,
 1267,
 5877,
 466,
 929,
 4281,
 1087,
 1332,
 684,
 2001,
 4557,
 4417,
 5354,
 2262,
 1803,
 4157,
 2271]

In [46]:
for i in y:
  if i in x:
    print(True)
  else:
    print(False)

True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True


In [48]:
user_activity = user_df[user_df["user_id"] == 1]
print(user_activity)


    id  user_id  recipe_id  action            timestamp
3    4        1       1219  search  2025-01-02 11:47:53
5    6        1       3088    view  2025-01-03 12:04:07
6    7        1       4889    like  2025-01-03 06:52:31
7    8        1        685    like  2025-01-03 08:08:57
8    9        1       1267    view  2025-01-02 19:20:13
11  12        1       5877  search  2025-01-02 22:30:54
14  15        1        466  search  2025-01-01 21:02:54
16  17        1        929    like  2025-01-02 06:24:24
22  23        1       4281  search  2025-01-02 22:04:18
25  26        1       1087    like  2025-01-02 06:52:57
27  28        1       1332    like  2025-01-02 04:13:28
28  29        1        684    like  2025-01-02 04:36:44
32  33        1       2001    view  2025-01-02 20:21:56
34  35        1       4557    view  2025-01-03 07:10:26
35  36        1       4417  search  2025-01-03 05:42:53
39  40        1       5354  search  2025-01-02 08:01:28
40  41        1       2262    like  2025-01-03 1

In [49]:
last_activity = user_activity.sort_values(by="timestamp", ascending=False).iloc[0]
print(last_activity)


id                             6
user_id                        1
recipe_id                   3088
action                      view
timestamp    2025-01-03 12:04:07
Name: 5, dtype: object


In [50]:
last_recipe_id = last_activity["recipe_id"]
print(last_recipe_id)

3088


In [51]:
user_recipe = recipe_df[recipe_df["recipe_id"] == last_recipe_id]
# print(user_recipe)
user_recipe

recipe_id       TranslatedRecipeName  \
3088       3088  Cardamom Macaroons Recipe   

                                  TranslatedIngredients  TotalTimeInMins  \
3088  2 Egg whites - aged for 2-3 days at room tempe...               70   

     Cuisine                             TranslatedInstructions  \
3088  French  To begin making the Cardamom Macaroons recipe,...   

                                    Cleaned-Ingredients  Ingredient-count  
3088  confectirs sugar,double,eggs aged days,food co...                10

In [57]:
user_vector = vectorizer.transform(user_recipe["Cleaned-Ingredients"])
print(user_vector)

  (0, 24)	0.3114252282543968
  (0, 34)	0.14634033045846184
  (0, 115)	0.14496747597188742
  (0, 139)	0.26432426269277515
  (0, 304)	0.10974930827594438
  (0, 339)	0.1089243332090492
  (0, 431)	0.17210959632360037
  (0, 482)	0.2611699691931175
  (0, 497)	0.27159887771631946
  (0, 583)	0.30099631973119484
  (0, 643)	0.2611699691931175
  (0, 655)	0.23456248086224718
  (0, 676)	0.12277214594336229
  (0, 682)	0.11054607675601195
  (0, 784)	0.25074106066991564
  (0, 1317)	0.18616668099653916
  (0, 1664)	0.09762213457954005
  (0, 1904)	0.27586869020055527
  (0, 2091)	0.18083039758337963
  (0, 2110)	0.3114252282543968
  (0, 2117)	0.14282742151716887
  (0, 2288)	0.1443602338586717


In [58]:
similarities = cosine_similarity(user_vector, ingredient_vectors)
print(similarities)

[[0.0324119  0.00567754 0.         ... 0.07498744 0.00759324 0.04843941]]


In [60]:
# Find similar recipes

similar_recipe_indices = similarities.argsort()[0][::-1][1:6]  # Exclude the current recipe
print(similar_recipe_indices)

[1409 3702 2936 2535 1873]


In [62]:
similarities #list of list

array([[0.0324119 , 0.00567754, 0.        , ..., 0.07498744, 0.00759324,
        0.04843941]])

This similarities contain the similer valuses from 0 to 1 to get this we need to sort it and get the most similer value

In [63]:
arr = np.array([5, 3, 1, 2, 4])
arr

array([5, 3, 1, 2, 4])

In [64]:
arr.argsort()

array([2, 3, 1, 4, 0])

In [65]:
# index in      5=0,  3=1 ,1=2 ,2=3, 4=4

# so in original array give this index ([5, 3, 1, 2, 4])
# in ascending  order place this value ([1 ,2 ,3, 4, 5 ])
# giving its original index            ([2 ,3 ,1 ,4 ,0])



In [66]:
similarities[0]

array([0.0324119 , 0.00567754, 0.        , ..., 0.07498744, 0.00759324,
       0.04843941])

In [67]:
similarities.argsort()[0]

array([3438, 5444, 4591, ..., 3702, 1409, 3088])

Using argsort we get the index values of that similarities

In [68]:
similarities.argsort()[0][::-1]   #[1:6]  # Exclude the current recipe


array([3088, 1409, 3702, ..., 4591, 5444, 3438])

Here [0][::-1] here we entering in to array as there is [[]] and then [::-1] means reverse this array so we will get the most similer value first ...like 3088 is exact match for that id then 1409

In [69]:
last_activity #most recent

id                             6
user_id                        1
recipe_id                   3088
action                      view
timestamp    2025-01-03 12:04:07
Name: 5, dtype: object

we got this from user id df  with decending value w.r.t time

In [70]:
user_recipe = recipe_df[recipe_df["recipe_id"] == last_recipe_id]
# print(user_recipe)
user_recipe.iloc[:, :2]

recipe_id       TranslatedRecipeName
3088       3088  Cardamom Macaroons Recipe

we got this from user id df ...user id df containing its search recipe id...we match that id with actual recipe df to get more information about that is..like recipe name ,ing etc

slicing to see id and name of that index recipe

In [71]:
similarities.argsort()[0][::-1]


array([3088, 1409, 3702, ..., 4591, 5444, 3438])

sorting similarity array  in ascending o to 1 and insted of value giving its index number in by argsort...y argmax we got values o to index...that means most similer will be at last as we want most simlier we want close to 1 so we reverse the array by[::-1]

In [72]:
similar_recipe_indices = similarities.argsort()[0][::-1][1:6]  # Exclude the current recipe ie present on first
similar_recipe_indices  #here list start from the same id or index that we already serach and which is exactly similer to our recipe we want similer of that 5 recommndation

array([1409, 3702, 2936, 2535, 1873])

after rversing we will get values 1 to 0 but we want most similer 5 value ...and the first index is the ame recipe which we search last time we want similer to that so we slicing 1 to 6 insted of 0 to 5

In [73]:
print(similar_recipe_indices)  # Print the indices of similar recipes

[1409 3702 2936 2535 1873]


In [74]:
def recommend_Xrecipes(selected_user):
    # Get user's last activity which contain user data w.r.t ids
    user_activity = user_df[user_df["user_id"] == selected_user] #(1,2,3)
    if user_activity.empty:
        return ["No history found for this user."]

    # Get the last searched or viewed recipe ID
    last_activity = user_activity.sort_values(by="timestamp", ascending=False).iloc[0]
    last_recipe_id = last_activity["recipe_id"]

    # Find the corresponding recipe in the recipe DataFrame
    if last_recipe_id not in recipe_df["recipe_id"].values:
        return ["Last activity recipe not found in the database."]

    user_recipe = recipe_df[recipe_df["recipe_id"] == last_recipe_id]

    # Find similar recipes
    user_vector = vectorizer.transform(user_recipe["Cleaned-Ingredients"])
    similarities = cosine_similarity(user_vector, ingredient_vectors)
    similar_recipe_indices = similarities.argsort()[0][::-1][1:6]  # Exclude the current recipe

    recommended_recipes = recipe_df.iloc[similar_recipe_indices]["TranslatedRecipeName"].tolist()
    return recommended_recipes

In [78]:
recommend_Xrecipes(selected_user=2)

['Awadhi Style Zarda Pulao Recipe',
 'Nariyal Ki Kheer Recipe - Coconut Pudding',
 'Saffron Pistachio Firni Recipe',
 'Kashmiri Pulao Recipe - Pulao Spiced With Saffron, Whole Spices & Nuts',
 'Basundi / Rabri Recipe']

#DEBUG

In [79]:
user_activity = user_df[user_df["user_id"] == 2]


In [80]:
user_activity.head(2)

id  user_id  recipe_id  action            timestamp
2   3        2       2438  search  2025-01-03 10:20:30
9  10        2       2281  search  2025-01-02 06:08:41

In [81]:
last_activity = user_activity.sort_values(by="timestamp", ascending=False).iloc[0]
#sorting df by colm =timestamp and in decending order so the bogger  value comes first
#and getting the fisrt most biggest value


In [82]:
last_activity

id                            21
user_id                        2
recipe_id                   5407
action                    search
timestamp    2025-01-03 11:16:53
Name: 20, dtype: object

In [83]:
serach_id=last_activity["recipe_id"]
serach_id

5407

In [84]:
#getting the recipe id
last_recipe_id = last_activity["recipe_id"]

In [85]:
last_recipe_id

5407

In [86]:
recipe_df.iloc[last_recipe_id][:2]

recipe_id                                                            5407
TranslatedRecipeName    Zarda Pulav Recipe (Rice Flavoured With Nuts A...
Name: 5407, dtype: object

In [87]:
# to get more info about this id we are puting this index in our original recipe data set
recipe_df.iloc[last_recipe_id]['Cleaned-Ingredients']

'black cardamom (badi elaichi),basmati rice,saffron strands generous,nutmeg powder,bay leaf (tej patta),ghee,cardamom (elaichi) podsseeds,pistachios,raisins,dry coconut (kopra),sugar,milk coconut milk almond milk,cinnamon stick (dalchini),badam (almond),cardamom powder (elaichi),cashews,cloves (laung)'

In [88]:
recommend_Xrecipes(selected_user=2)

['Awadhi Style Zarda Pulao Recipe',
 'Nariyal Ki Kheer Recipe - Coconut Pudding',
 'Saffron Pistachio Firni Recipe',
 'Kashmiri Pulao Recipe - Pulao Spiced With Saffron, Whole Spices & Nuts',
 'Basundi / Rabri Recipe']

In [89]:
import joblib
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer()
ingredient_vectors = vectorizer.fit_transform(recipe_df["Cleaned-Ingredients"])





In [90]:
# Save the vectorizer and ingredient vectors
joblib.dump(vectorizer, 'tfidf_vectorizer.pkl')
joblib.dump(ingredient_vectors, 'ingredient_vectors.pkl')

['ingredient_vectors.pkl']